### PR TITLE
Fix DiLoCo training compatibility issues

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

Summary:
  - Fix mesh discovery for drjax: Pass mesh explicitly to all drjax.broadcast/drjax.map_fn calls since jax.set_mesh() uses a different thread-local than pxla.thread_resources (which drjax reads). Also add with mesh: alongside jax.set_mesh() in train and train_compile
  entrypoints.
  - Fix data loading order: Move jax.device_put after reshape_first_axis_with_diloco in DataLoader.load_next_batch, and add the diloco reshape to RampUpDataLoader as well. Guard sharding access in reshape_for_diloco for arrays without a .sharding attribute.
  - Fix outer optimizer state sharding: Shard the SGD momentum trace like model params instead of fully replicating it (replicated sharding caused issues).
  - Auto-resolve dcn_diloco_parallelism=-1: Infer the diloco DCN parallelism from num_slices and other DCN axes, matching the existing convention for dcn_data_parallelism.
  - Add use_tokamax_gmm + enable_diloco incompatibility check: Raise a clear error since tokamax's GroupSizes vmap_rule breaks under drjax's jax.vmap batching.
  - Add diloco mesh axis to the deepseek3-671b-2dfsdp config.
  

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/types.py [hunk 1]
```diff
@@ -2650,8 +2643,32 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
...
+          f"Cannot resolve dcn_diloco_parallelism=-1: num_slices={self.num_slices} is not divisible "
+          f"by the product of other DCN parallelism values ({other_dcn_product})."
+      )
+      self.dcn_diloco_parallelism = self.num_slices // other_dcn_product
+      # Keep dcn_parallelism list consistent with the resolved value.
+      diloco_idx = self.dcn_parallelism.index(-1)
+      self.dcn_parallelism[diloco_idx] = self.dcn_diloco_parallelism
     self.num_diloco_replicas = int(self.ici_diloco_parallelism * self.dcn_diloco_parallelism)
 
+    # use_tokamax_gmm is incompatible with enable_diloco: drjax.map_fn wraps the train step in
```
Comment (gobbleturk) [inline src/maxtext/configs/types.py]: can we create a bug for this to track?

File: src/maxtext/configs/types.py [hunk 2]
```diff
@@ -2650,8 +2643,32 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
...
+    # jax.vmap over the diloco axis, which causes JAX to batch through lax.scan (layer scan).
+    # Tokamax's vmap_rule then tries to reconstruct GroupSizes with a batched 2-D value, but
+    # GroupSizes.__post_init__ requires exactly a 1-D shape.
+    if self.enable_diloco and self.use_tokamax_gmm:
+      raise ValueError(
+          "use_tokamax_gmm=True is not compatible with enable_diloco=True due to a known "
+          "incompatibility between tokamax's GroupSizes vmap_rule and JAX's scan batching. "
+          "Please set use_tokamax_gmm=False."
+      )
+
```
Comment (NuojCheng) [inline src/maxtext/configs/types.py:2708]: I wonder if diloco works with all opt_type? e.g. sgd, muon. Maybe disable them if they don't work.

File: src/maxtext/trainers/pre_train/train.py [hunk 3]
```diff
@@ -502,19 +490,18 @@ def train_loop(config, recorder, state=None):
...
+        config,
+        model,
+        mesh,
+        state,
+        state_mesh_shardings,
+        train_step,
+        eval_step,
+        eval_data_iterator,
+        params_shardings,
+    )
```
Comment (NuojCheng) [inline src/maxtext/trainers/pre_train/train.py:521]: our jit function was not in the mesh/logical rule context before? what difference did you see by making this change?

Comment (NuojCheng) [reply src/maxtext/trainers/pre_train/train.py:521]: oh I see the comments in train_compile. I think JAX will deprecate with mesh context.

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (khatwanimohit) [reply src/maxtext/configs/types.py]: done

File: src/maxtext/configs/types.py (hunk 2 above)
Comment (khatwanimohit) [reply src/maxtext/configs/types.py:2708]: it works with all opt_types for internal optimizer, the outer optimizer is hard coded to nesterov momentum sgd
